rlp/encode.go

```diff
@@ -397,22 +399,93 @@ func writeBigIntNoPtr(val reflect.Value, w *encbuf) error {
 }
 
 func writeBigInt(i *big.Int, w *encbuf) error {
-	if cmp := i.Cmp(big0); cmp == -1 {
+	switch i.Sign() {
+	case -1:
 		return fmt.Errorf("rlp: cannot encode negative *big.Int")
-	} else if cmp == 0 {
+	case 0:
 		w.str = append(w.str, 0x80)
-	} else {
-		w.encodeString(i.Bytes())
+	default:
+		bitlen := i.BitLen()
+		if bitlen <= 64 {
+			w.encodeUint(uint64(i.Bits()[0]))
```

Comment:
This won't work on 32 bit architecture, right? Because then the first word is only 32 bits, not 64.

Reply:
`uint64(i.Int64())` should do the trick right?